storage/response_cache.go

```diff
@@ -72,26 +72,32 @@ func (rc *ResponseCache) ClearData(e engine.Engine) error {
 	return err
 }
 
-// GetResponse looks up a response matching the specified cmdID and
-// returns true if found. The response is deserialized into the
-// supplied reply parameter. If no response is found, returns
-// false. If a command is pending already for the cmdID, then this
-// method will block until the the command is completed or the
-// response cache is cleared.
-func (rc *ResponseCache) GetResponse(e engine.Engine, cmdID proto.ClientCmdID) (proto.Response, error) {
+// GetResponse looks up a response matching the specified cmdID. If the
+// response is found, it is returned along with its associated error.
+// If the response is not found, nil is returned for both the response
+// and its error. In all cases, the third return value is a
+// *replicaCorruptionError that is non-nil when the response cache is
+// unable to read from storage.
+func (rc *ResponseCache) GetResponse(e engine.Engine, cmdID proto.ClientCmdID) (proto.Response, error, error) {
 	// Do nothing if command ID is empty.
 	if cmdID.IsEmpty() {
-		return nil, nil
+		return nil, nil, nil
 	}
 
 	// Pull response from the cache and read into reply if available.
-	rwResp := proto.ReadWriteCmdResponse{}
+	var rwResp proto.ReadWriteCmdResponse
 	key := keys.ResponseCacheKey(rc.raftID, &cmdID)
 	ok, err := engine.MVCCGetProto(e, key, proto.ZeroTimestamp, true, nil, &rwResp)
-	if ok && err == nil {
-		return rwResp.GetValue().(proto.Response), nil
+	if err != nil {
+		return nil, nil, newReplicaCorruptionError(util.Errorf("could not read from response cache"), err)
```

Comment:
I've actually thought about putting ResponseCache in its own package instead of making it private to `storage`. I'm not sure it's big enough to give its own package but it's reasonably self-contained and so it's one thing we could do to shrink the storage monolith.

If we did that, I'd argue that replicaCorruptionError should be on the `Range` side of the split and not local to the `ResponseCache`. The responseCache shouldn't know anything about how its failings impact replication.